File: programmers/old/kakao/뉴스클러스터링.java
package A1;

import java.io.*;
import java.util.*;

class Solution {
    public int solution(String str1, String str2) {
        str1 = str1.toLowerCase();
        str2 = str2.toLowerCase();
        Map<String, Integer> map1 = new HashMap<String,Integer>();
        Map<String, Integer> map2 = new HashMap<String,Integer>();
        for(int i = 0; i < str1.length()-1; ++i) {
        	if(str1.charAt(i) >= 'a' && str1.charAt(i) <= 'z' && str1.charAt(i+1) >= 'a' && str1.charAt(i+1) <= 'z') {
        		String s = str1.substring(i,i+2);
        		if(map1.containsKey(s)) {
        			map1.replace(s, map1.get(s)+1);
        		}else {
        			map1.put(s,1);
        		}
        	}
        }
        for(int i = 0; i < str2.length()-1; ++i) {
        	if(str2.charAt(i) >= 'a' && str2.charAt(i) <= 'z' && str2.charAt(i+1) >= 'a' && str2.charAt(i+1) <= 'z') {
        		String s = str2.substring(i,i+2);
        		if(map2.containsKey(s)) {
        			map2.replace(s, map2.get(s)+1);
        		}else {
        			map2.put(s,1);
        		}
        	}
        }
        Map<String, Integer> intersection = new HashMap<>();
        Map<String,Integer> union = new HashMap<>();
        for(String s1: map1.keySet()) {
        	union.put(s1, map1.get(s1));
        	for(String s2: map2.keySet()) {
        		if(s1.equals(s2)) {
        			intersection.put(s1, Math.min(map1.get(s1),  map2.get(s2)));
        		}
        	}
        }
        for(String s2: map2.keySet()) {
        	if(union.containsKey(s2)) {
        		union.replace(s2,  Math.max(union.get(s2),  map2.get(s2)));
        	}else {
        		union.put(s2, map2.get(s2));
        	}
        }
        int interSum = 0;
        int unionSum = 0;
        for(int num : intersection.values()) {
        	interSum += num;
        }
        for(int num : union.values()) {
        	unionSum += num;
        }
        int answer = 65536;
        if(unionSum > 0) {
        	answer = interSum * 65536 / unionSum;
        }
        return answer;
    }
}

public class 뉴스클러스터링 {
	public static void main(String[] args) {
		String str1 = "FRANCE";
		String str2 = "french";
		Solution sol = new Solution();
		System.out.println(sol.solution(str1, str2));
	}
}

File: programmers/js/new/level2/뉴스_클러스터링.js
function solution(str1, str2) {
    str1 = str1.toLowerCase();
    str2 = str2.toLowerCase();
    const map1 = new Map();
    for (let i = 0; i < str1.length; ++i) {
        const tStr = str1.substring(i, i + 2);
        // 정규식으로 해보기
        if (
            ((tStr[0] >= "a" && tStr[0] <= "z") || (tStr[0] >= "A" && tStr[0] <= "Z")) &&
            ((tStr[1] >= "a" && tStr[1] <= "z") || (tStr[1] >= "A" && tStr[1] <= "Z"))
        ) {
            if (map1.has(tStr)) {
                map1.set(tStr, map1.get(tStr) + 1);
            } else {
                map1.set(tStr, 1);
            }
        }
    }
    const map2 = new Map();
    for (let i = 0; i < str2.length; ++i) {
        const tStr = str2.substring(i, i + 2);
        // 정규식으로 해보기
        if (
            ((tStr[0] >= "a" && tStr[0] <= "z") || (tStr[0] >= "A" && tStr[0] <= "Z")) &&
            ((tStr[1] >= "a" && tStr[1] <= "z") || (tStr[1] >= "A" && tStr[1] <= "Z"))
        ) {
            if (map2.has(tStr)) {
                map2.set(tStr, map2.get(tStr) + 1);
            } else {
                map2.set(tStr, 1);
            }
        }
    }
    return parseInt((65536 * intersect(map1, map2)) / union(map1, map2));
}

function intersect(map1, map2) {
    if (map1.size === 0 && map2.size === 0) {
        return 1;
    }
    let count = 0;
    map1.forEach((v, k) => {
        if (map2.has(k)) {
            const v2 = map2.get(k);
            count += Math.min(v, v2);
        }
    });
    return count;
}

function union(map1, map2) {
    if (map1.size === 0 && map2.size === 0) {
        return 1;
    }
    let count = 0;
    map1.forEach((v, k) => {
        if (map2.has(k)) {
            const v2 = map2.get(k);
            count += Math.max(v, v2);
        } else {
            count += v;
        }
    });
    map2.forEach((v, k) => {
        if (!map1.has(k)) {
            count += v;
        }
    });
    return count;
}

console.log(solution("E=M*C^2", "e=m*c^2"));
